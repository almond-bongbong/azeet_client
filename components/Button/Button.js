import React, { useMemo } from 'react';
import PropTypes from 'prop-types';
import SimpleLoader from 'components/Loader/SimpleLoader';
import styled, { css } from 'styled-components';
import { red, blue } from 'style/mixin';

const BasicButtonStyle = styled.div`
  display: block;
  
  ${props => props.inline && css`
    display: inline-block;
    width: auto;
  `};

  button, a {
    display: block;
    width: ${props => `${props.width !== 'auto' ? `${props.width}px` : '100%'}`};
    min-width: 80px;
    height: ${props => props.height}px;
    padding: 0 20px;
    border: 1px solid #ddd;
    border-radius: 2px;
    outline: 0;
    font-size: 13px;
    cursor: pointer;
    text-align: center;
    text-decoration: none;
    line-height: ${props => props.height}px;
    transition: box-shadow .2s;
  
    &:hover {
      box-shadow: 1px 1px 3px rgba(0,0,0,0.2);
    }
    
    &:active {
      box-shadow: 0 0 2px rgba(0,0,0,0.1);
    }
    
    & span {
      position: relative;
    }
    
    ${props => props.theme === 'white' && css`
      background-color: #fff;
      border: 1px solid #ddd;
      color: #666;
    `};
    
    ${props => props.theme === 'gray' && css`
      background-color: #666;
      border: 1px solid #666;
      color: #fff;
    `}
    
    ${props => props.theme === 'red' && css`
      background-color: ${red};
      border: 1px solid ${red};
      color: #fff;
    `};
    
    ${props => props.theme === 'blue' && css`
      background-color: ${blue};
      border: 1px solid ${blue};
      color: #fff;
    `}
  }
`;

const Button = ({
  text, theme, width, height, inline, loading, onClick,
}) => {
  const loaderTheme = useMemo(() => (theme === 'white' ? 'red' : 'white'), [theme]);
  const loaderSize = useMemo(() => height / 2, [height]);

  return (
    <BasicButtonStyle data-testid="button-wrap" theme={theme} width={width} height={height} inline={inline}>
      {loading ? (
        <button type="button" title="로딩중"><SimpleLoader theme={loaderTheme} size={loaderSize} /></button>
      ) : (
        <button type="button" onClick={onClick}><span>{text}</span></button>
      )}
    </BasicButtonStyle>
  );
};

Button.propTypes = {
  text: PropTypes.string,
  inline: PropTypes.bool,
  theme: PropTypes.oneOf(['white', 'gray', 'red', 'blue']),
  width: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  height: PropTypes.oneOf([30, 40, 50]),
  loading: PropTypes.bool,
  onClick: PropTypes.func,
};

Button.defaultProps = {
  text: '버튼',
  theme: 'white',
  width: 'auto',
  height: 40,
  inline: true,
  loading: false,
  onClick: undefined,
};

export default Button;
