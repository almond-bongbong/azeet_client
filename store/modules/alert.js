import { handleActions, createAction } from 'redux-actions';
import produce from 'immer';

export const alertTypes = {
  OPEN: 'alert/OPEN',
  CLOSE: 'alert/CLOSE',
};

export const alertActions = {
  open: createAction(alertTypes.OPEN),
  close: createAction(alertTypes.CLOSE),
  alert: message => dispatch => new Promise((resolve) => {
    dispatch(alertActions.open({ message, resolve }));
  }),
};

const initialState = {
  show: false,
  message: null,
};

export default handleActions({
  [alertTypes.OPEN]: (state, { payload }) => produce(state, (draft) => {
    draft.show = true;
    draft.message = payload.message;
  }),
  [alertTypes.CLOSE]: state => produce(state, (draft) => {
    draft.show = false;
  }),
}, initialState);
