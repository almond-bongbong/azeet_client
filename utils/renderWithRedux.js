import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import configStore from 'store/configure';
import Alert from 'components/Alert';
import Confirm from 'components/Confirm';
import Toast from 'components/Toast';

function renderWithRedux(ui, initialState) {
  const store = configStore(initialState);
  const utils = render(
    <Provider store={store}>
      {ui}
      <Alert />
      <Confirm />
      <Toast />
    </Provider>,
  );
  return {
    ...utils,
    store,
  };
}

export default renderWithRedux;
