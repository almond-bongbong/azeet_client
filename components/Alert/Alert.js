import React from 'react';
import { useSelector } from 'react-redux';
import ReactCSSTransitionGroup from 'react-addons-css-transition-group';
import styled from 'styled-components';
import Button from 'components/Button/Button';
import newlineText from 'lib/newlineText';
import { fadeIn } from 'style/animations';

const AlertStyle = styled.div`
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: 10000;
  background-color: rgba(0,0,0,0.4);
  
  &.alert-enter {
    opacity: 0;
    transition: opacity .2s;
  }
  &.alert-enter-active {
    opacity: 1;
  }
  &.alert-leave {
    opacity: 1;
    transition: opacity .2s;
  }
  &.alert-leave-active {
    opacity: 0;
  }
  
  & .content {
    position: absolute;
    top: 50%;
    left: 50%;
    width: 250px;
    padding: 20px;
    border-radius: 6px;
    background-color: #fff;
    box-shadow: 1px 1px 10px 1px rgba(0,0,0,0.2);
    text-align: center;
    transform: translate(-50%, -50%);
  }
  & .message {
    margin-bottom: 25px;
    color: #666;
    font-weight: 700;
    font-size: 16px;
  }
`;

const Alert = () => {
  const { show, message } = useSelector(state => state.alert);

  return (
    <ReactCSSTransitionGroup
      transitionName="alert"
      transitionEnterTimeout={200}
      transitionLeaveTimeout={200}
    >
      {show && (
        <AlertStyle>
          <div className="content">
            <div className="message">
              {newlineText(message)}
            </div>
            <Button inline={false} onClick={window.callbackStore.alert} text="확인" />
          </div>
        </AlertStyle>
      )}
    </ReactCSSTransitionGroup>
  );
};

export default Alert;
