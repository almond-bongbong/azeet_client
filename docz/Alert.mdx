---
name: Alert
menu: Feedback
---

import { Playground } from 'docz'
import { useDispatch } from 'react-redux'
import { alert } from 'components/Feedback'
import Button from 'components/Button/Button'

# alert

## Basic usage

- 아래와 같은 형식으로 alert 팝업을 호출 할 수 있습니다.
```
const openAlert = () => alert('안녕하세요')
```

- alert는 `Promise` 객체를 반환 하므로 팝업이 종료되는 시점을 확인할 수 있습니다.
```
const openAlert = () => alert('하이').then(() => console.log('close'))
```

<Playground>
  {() => {
    const Test = () => {
      const openAlert = () => alert('하이').then(() => console.log('close'))
      return <Button onClick={openAlert}>alert!</Button>
    }

    return <Test />
  }}
</Playground>
