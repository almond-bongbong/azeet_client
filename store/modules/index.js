import { combineReducers } from 'redux';
import alert from './alert';
import confirm from './confirm';
import toast from './toast';
import auth from './auth';
import bookmark from './bookmark';
import pending from './pending';

export default combineReducers({
  alert,
  confirm,
  toast,
  auth,
  bookmark,
  pending,
});
