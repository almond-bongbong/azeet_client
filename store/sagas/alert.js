import { eventChannel, END } from 'redux-saga';
import { all, fork, call, put, take, takeLatest } from 'redux-saga/effects';
import { alertActions, types } from 'store/modules/alert';

const alertChannel = () => eventChannel((emit) => {
  window.callbackStore = { ...window.callbackStore, alert: () => emit(true) };
  return () => emit(END);
});

function* alert({ payload }) {
  const channel = yield call(alertChannel);
  yield take(channel);
  yield call(payload.resolve);
  yield put(alertActions.close());
}

function* watchAlert() {
  yield takeLatest(types.OPEN, alert);
}

export default function* alertSaga() {
  yield all([
    fork(watchAlert),
  ]);
}
