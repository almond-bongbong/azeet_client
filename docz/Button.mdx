---
name: Button
---

import { Playground, Props } from 'docz'
import Button from 'components/Button/Button'

# Button

<Props of={Button} />

## basic usage

<Playground>
  <Button>Button</Button>
</Playground>

## inline

### true

<Playground>
  <Button theme="red">버튼</Button>
</Playground>

### false

<Playground>
  <Button theme="red" inline={false}>버튼</Button>
</Playground>

## theme

### white

<Playground>
  <Button theme="white">취소</Button>
</Playground>

### gray

<Playground>
  <Button theme="gray">취소</Button>
</Playground>

### red

<Playground>
  <Button theme="red">버튼</Button>
</Playground>

### blue

<Playground>
  <Button theme="blue">버튼</Button>
</Playground>

## width

<Playground>
  <Button width={120} theme="white">작은거</Button>
</Playground>

## height

### 30

<Playground>
  <Button height={30} theme="white">작은거</Button>
</Playground>

### 40

<Playground>
  <Button height={40} theme="white">보통</Button>
</Playground>

### 50

<Playground>
  <Button height={50} theme="white">큰거</Button>
</Playground>

## loading

<Playground>
  <Button loading={true}>로딩</Button>
</Playground>

## event

### onClick(event)

<Playground>
  <Button theme="white" onClick={e => alert('클릭')}>클릭!</Button>
</Playground>