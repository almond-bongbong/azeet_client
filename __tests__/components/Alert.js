import React from 'react';
import { fireEvent, wait } from '@testing-library/react';
import renderWithRedux from 'utils/renderWithRedux';
import Alert from 'components/Alert/Alert';
import Button from 'components/Button/Button';
import { alert } from 'store/modules/alert';

describe('Alert', () => {
  it('matches snapshot', () => {
    const utils = renderWithRedux(<Alert />);
    expect(utils.container).toMatchSnapshot();
  });

  it('calls alert and close', async () => {
    const { getByText, queryByText, store } = renderWithRedux(
      <>
        <Button text="click" onClick={() => store.dispatch(alert('alert'))} />
        <Alert />
      </>,
    );

    const button = getByText('click');
    fireEvent.click(button);
    getByText('alert');

    const closeButton = getByText('확인');
    fireEvent.click(closeButton);
    await wait(() => expect(queryByText('alert')).toBeNull());
  });
});
