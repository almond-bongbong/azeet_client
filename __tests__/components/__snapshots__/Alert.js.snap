// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert matches snapshot 1`] = `
<div>
  <span />
  <span />
  <span />
  <div
    class="Toast__ToastStyle-sc-18p3u8j-0 jhiKCx"
  >
    <span />
  </div>
</div>
`;
